packages/gatsby/src/db/loki/index.js

```diff
@@ -0,0 +1,120 @@
+const _ = require(`lodash`)
+const fs = require(`fs-extra`)
+const path = require(`path`)
+const loki = require(`@moocar/lokijs`)
+const uuidv4 = require(`uuid/v4`)
+
+// Loki is a document store with the same semantics as mongo. This
+// means there are no tables or relationships. Just a bunch of
+// collections, each with objects.
+//
+// Gatsby stores nodes in collections by splitting them up by their
+// `node.internal.type`. All nodes of a particular type go in 1
+// collection. The below `colls` object contains the metadata for
+// these collections, and the "meta collections" used to track them.
+//
+// You won't use these directly. They are used by the collection
+// functions in `./nodes.js`. E.g `getTypeCollName()` and
+// `getNodeTypeCollection`
+const colls = {
+  // Each object has keys `id` and `typeCollName`. It's a way of
+  // quickly looking up the collection that a node is contained in.
+  // E.g { id: `someNodeId`, typeCollName: `gatsby:nodeType:myType` }
+  nodeMeta: {
+    name: `gatsby:nodeMeta`,
+    options: {
+      unique: [`id`],
+      indices: [`id`],
+    },
+  },
+  // The list of all node type collections. Each object has keys
+  // `type` and `collName` so you can quickly look up the collection
+  // name for a node type.
+  // e.g { type: `myType`, collName: `gatsby:nodeType:myType` }
+  nodeTypes: {
+    name: `gatsby:nodeTypes`,
+    options: {
+      unique: [`type`, `collName`],
+      indices: [`type`],
+    },
+  },
+}
+
+// Must be set using `start()`
+let db
+
+/**
+ * Ensures that the collections that support nodes have been
+ * created. See `colls` var in this file
+ */
+function ensureNodeCollections(db) {
+  _.forEach(colls, collInfo => {
+    const { name, options } = collInfo
+    db.addCollection(name, options)
+  })
+}
+
+function startFileDb(saveFile) {
+  return new Promise((resolve, reject) => {
+    const dbOptions = {
+      autoload: true,
+      autoloadCallback: err => {
+        if (err) {
+          reject(err)
+        } else {
+          resolve()
+        }
+      },
+      autosave: true,
+      autosaveInterval: 1000,
+    }
+    db = new loki(saveFile, dbOptions)
+  })
+}
+
+async function startInMemory() {
+  // Use uuid purely for a random name
+  db = new loki(uuidv4())
+}
```

Comment:
this is just for unit tests?

Reply:
Yeah, normally I would add this in a `__tests__` file, but it's so small and easier to use in a production namespace. I can move it out of prod code if you want though.